### ISSUE TITLE
[DTensor][local_map]Support `None` placements in `local_map` placement specs

### ISSUE THREAD

**Issue by wwwjn** [pytorch/pytorch#186030]
## Feature Request

`torch.distributed.tensor.experimental.local_map` should support placement specs that contain
`None` for individual input/output/gradient placements, now it only supports a tuple or None, not [NamedPlacementA, NamedPlacementB, None]. 
This assertion errors out when here's `None` placements within the tuple: https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/experimental/_func_map.py#L218

## Motivation

Some module signatures naturally contain mixed tensor and non-tensor values, or optional tensor values. For those cases, we want to describe only the tensor leaves that should be converted by `local_map`, while preserving `None` leaves as pass-through values.

For example, a function may have:

```python
def fn(x, maybe_aux):
    ...
    return y, maybe_aux

where x and y are DTensor-backed values, but maybe_aux can be None.

Ideally, local_map could be configured as:

wrapped = local_map(
    fn,
    device_mesh=device_mesh,
    in_placements=((Shard(0),), None),
    out_placements=((Shard(0),), None),
    in_grad_placements=((Shard(0),), None),
)

```
and the None placement leaves would mean:

- Do not convert this input/output tensor.
- Do not expect a DTensor placement for that parameter.
- Do not participate in gradient placement handling for that placement.

I want to discuss if there's something I failed to consider and why we not allow None in the groups. cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @tianyu-l @XilunWu @SherlockNoMad @ppwwyyxx

**Comment by tianyu-l** [comment]
cc @pianpwk

**Comment by RohitRathore1** [comment]
Hi @wwwjn, this feature request makes sense to me the current assertion blocks a natural use case for mixed-signature functions. I'd like to work on this. I am thinking

- when spec is None for a DTensor input, skip redistribution and pass the DTensor through unchanged (no to_local() conversion).
- when spec is None for a DTensor output, skip the DTensor.from_local() wrapping and pass the tensor through as-is.
- and, when spec is None for grad placements, call to_local() without grad_placements.

before I start, I wanted to confirm should None mean "pass through the DTensor as-is" or "call to_local() with default placements"? The former seems more useful for the motivating use case.
cc: @pianpwk @tianyu-l

**Comment by aditvenk** [comment]
cc @xmfan could you weigh in ?

**Comment by wconstab** [comment]
@wwwjn one clarification: were you proposing to support passing in DTensors along with a None placement, which tells local map to just pass that DTensor placement through?  Or you just want to allow passing an actual None (or non tensor) input, along with a None placement?  The latter makes sense to me, i'm not sure about the former.

**Comment by wwwjn** [comment]
@wconstab I was proposing the latter, passing a non-tensor input along with None placement

**Comment by RohitRathore1** [comment]
Thanks for the clarification @wwwjn. I want to make sure I'm scoping this correctly before starting. Looking at the current code, non-tensor inputs with None placement already pass through correctly, the else branch at https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/experimental/_func_map.py#L232-L242 handles exactly that case for inputs, and https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/experimental/_func_map.py#L274-L280 handles it for outputs. The assertion you originally referenced only fires when the arg is a DTensor and its grad placement spec is None. Similarly, https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/experimental/_func_map.py#L195-L198 blocks DTensor inputs with None in in_placements. Could you share a minimal reproducer that triggers the error? That would help me confirm which code path actually needs to change. It's possible the issue is subtler than it appears (e.g., a pytree flattening mismatch between args and placements).

cc: @wconstab

**Comment by wconstab** [comment]
i agree with @wwwjn that limiting scope to support None placements accompanying non-DTensor inputs is plenty, we should probably ban DTensor inputs + None placements.

Given that, if you're unable to repro the original code snippet @wwwjn pasted, i think it's fair to stop until @wwwjn can provide another repro.  You might also want to ask an agent whether it can find a similar repro that currently fails and then go fix that.

**Comment by RohitRathore1** [comment]
Thanks @wconstab. I ran a comprehensive reproducer testing all combinations of `None` in placement specs, and here are the results:

**Option B (non-tensor input + `None` placement) already works.** The [`else` branch at L232-L242](https://github.com/pytorch/pytorch/blob/aab577b23644/torch/distributed/tensor/experimental/_func_map.py#L232-L242) handles non-tensor inputs with `None` placement correctly, and [L274-L280](https://github.com/pytorch/pytorch/blob/aab577b23644/torch/distributed/tensor/experimental/_func_map.py#L274-L280) handles non-tensor outputs. There's existing test coverage in:
- [`test_local_map.py` Test 3 (L219-L234)](https://github.com/pytorch/pytorch/blob/aab577b23644/test/distributed/tensor/experimental/test_local_map.py#L219-L234) — scalar with `None` placement
- [`test_local_map.py` `test_none_placements` (L751-L786)](https://github.com/pytorch/pytorch/blob/aab577b23644/test/higher_order_ops/test_local_map.py#L751-L786) — scalar and module with `None` placement

**The assertions the issue originally referenced only fire for DTensor inputs (Option A):**
- [L195-L198](https://github.com/pytorch/pytorch/blob/aab577b23644/torch/distributed/tensor/experimental/_func_map.py#L195-L198) — blocks DTensor input with `None` in `in_placements`
- [L218-L221](https://github.com/pytorch/pytorch/blob/aab577b23644/torch/distributed/tensor/experimental/_func_map.py#L218-L221) — blocks DTensor input with `None` in `in_grad_placements`

**One related issue I did find:** at [L270-L273](https://github.com/pytorch/pytorch/blob/aab577b23644/torch/distributed/tensor/experimental/_func_map.py#L270-L273), if a function returns a regular `torch.Tensor` output with `spec=None`, it gets silently wrapped as a replicated DTensor via `DTensor.from_local(out, mesh, None)` instead of being passed through unchanged. This could bite users who return mixed `(tensor, regular_tensor)` outputs.

@wwwjn could you share a minimal repro that triggers the assertion? The exact example from the issue body (non-tensor `None` input with `None` placement) passes without error on current `main`.

**Comment by wwwjn** [comment]
Thanks @RohitRathore1 and @wconstab. You are correct that the non-tensor example in the issue body already works. I traced the original torchtitan failure, and the issue body conflated two different cases.

The actual use case is a DTensor input with:
- a valid `in_placements` entry describing its forward layout
- `None` for only its corresponding `in_grad_placements` entry

For example:

```python
shard = (Shard(0),)
replicate = (Replicate(),)

x_dt = distribute_tensor(x, mesh, shard)
metadata_dt = distribute_tensor(metadata, mesh, replicate)

def fn(x, metadata):
    return x

wrapped = local_map(
    fn,
    device_mesh=mesh,
    in_placements=(shard, replicate),
    in_grad_placements=(shard, None),
    out_placements=shard,
)

wrapped(x_dt, metadata_dt)
```

The motivating TorchTitan input was `num_local_tokens_per_expert_E` (a DTensor metadata for MoE routing, which don't need gradient). TorchTitan worked around the API restriction by supplying the tensor's existing placement as an otherwise-unused gradient-placement hint: https://github.com/pytorch/torchtitan/blob/388a6062a/torchtitan/models/common/moe_sharding.py#L275-L283. The workaround is numerically correct however it's  redundant because the metadata tensor has `requires_grad=False` already, so putting a grad placement is redundant. 

After correcting my understanding, do you think we should still consider supporting this use case?
